reader.go

```diff
@@ -284,39 +290,65 @@ func (r *reader) run(ctx context.Context, offset int64) {
 		// to the connection we know we'll want to restart from this offset.
 		offset = start
 
+		errcount := 0
 	readLoop:
 		for {
+			if !sleep(ctx, backoff(errcount, 100*time.Millisecond, 10*time.Second)) {
+				conn.Close()
+				return
+			}
+
 			switch offset, err = r.read(ctx, offset, conn); err {
 			case nil:
+				errcount = 0
+				continue
 			case RequestTimedOut:
-			case context.Canceled: // another reader has taken over
+				// Timeout on the kafka side, this can be safely retried.
+			case OffsetOutOfRange:
```

Comment:
That's a good point, although if the messages are gone they're already lost and I feel like we're better off trying to consume whatever is available. I do agree we should have a good story around being notified of those things, but getting an offset-out-of-range error doesn't communicate this well.